test/dlc_tests/conftest.py

```diff
@@ -222,10 +236,210 @@ def pytest_runtest_setup(item):
+        write_test_coverage_file(test_cov, test_coverage_file)
+
+        if failure_conditions:
+            message, total_issues = _assemble_report_failure_message(failure_conditions)
+            if total_issues == 0:
+                LOGGER.warning(f"Found failure message, but no issues. Message:\n{message}")
+            else:
+                raise TestReportGenerationFailure(message)
+
+
+def _handle_single_gpu_instances(function_key, function_keywords, failures, processor="gpu"):
+    """
+    Generally, we do not want tests running on single gpu instance types. However, there are exceptions to this rule.
+    This function is used to determine whether we need to raise an error with report generation or not, based on
+    whether we are using single gpu instances or not in a given test function.
+
+    :param function_key: local/path/to/function::function_name
+    :param function_keywords: string of keywords associated with the test function
+    :param failures: running dictionary of failures associated with the github link
+    :param processor: whether the test is for cpu, gpu or both
+    :return: processor if not single gpu instance, else "single_gpu", and a dict with updated failure messages
+    """
+
+    # Define conditions where we allow a test function to run with a single gpu instance
+    whitelist_single_gpu = False
+    allowed_single_gpu = ("telemetry", "test_framework_version_gpu")
+
+    # Regex in order to determine the gpu instance type
+    gpu_instance_pattern = re.compile(r"\w+\.\d*xlarge")
+    gpu_match = gpu_instance_pattern.search(function_keywords)
+
+    if gpu_match:
+        instance_type = gpu_match.group()
+        num_gpus = ec2_utils.get_instance_num_gpus(instance_type=instance_type)
+
+        for test in allowed_single_gpu:
+            if test in function_key:
+                whitelist_single_gpu = True
+                break
+        if num_gpus == 1:
+            processor = "single_gpu"
+            if not whitelist_single_gpu:
+                single_gpu_failure_message = (
+                    f"Function {function_key} uses single-gpu instance type {instance_type}. "
+                    f"Please use multi-gpu instance type."
+                )
+                if not failures.get(function_key):
+                    failures[function_key] = [single_gpu_failure_message]
+                else:
+                    failures[function_key].append(single_gpu_failure_message)
+
+    return processor, failures
+
+
+def _assemble_report_failure_message(failure_messages):
+    """
+    Function to assemble the failure message if there are any to raise
+
+    :param failure_messages: dict where key is the function, and value is a list of failures associated with the
+    function
+    :return: the final failure message string
+    """
+    final_message = ""
+    total_issues = 0
+    for func, messages in failure_messages.items():
+        final_message += f"******Problems with {func}:******\n"
```

Comment:
This line seems redundant based on the error message shared in the description.

Reply:
The idea here is that one can add errors to a given function via failure_messages. So if there were errors to add outside of just single gpu, they would be displayed under the function that they affect

Reply:
Correct - the error message for multi_gpu also includes {function_key} - that can be removed as the function name is printed by this function anyway.

Reply:
oh i see, sure i can remove it